core/modules/memman/MemManReal.cc

```diff
@@ -0,0 +1,231 @@
+#include "memman/MemFile.h"
+#include "memman/MemFileSet.h"
+
+/******************************************************************************/
+/*                  L o c a l   S t a t i c   O b j e c t s                   */
+/******************************************************************************/
+  
+namespace {
+
+std::mutex hanMutex;
+
+std::unordered_map<lsst::qserv::memman::MemMan::Handle,
+                   lsst::qserv::memman::MemFileSet*> hanCache;
+
+lsst::qserv::memman::MemMan::Handle handleNum
+                             = lsst::qserv::memman::MemMan::HandleType::ISEMPTY;
+}
+
+namespace lsst {
+namespace qserv {
+namespace memman {
+  
+/******************************************************************************/
+/*                         g e t S t a t i s t i c s                          */
+/******************************************************************************/
+  
+MemMan::Statistics MemManReal::getStatistics() {
+
+Statistics stats;
+
+    // Get all the needed information and return it
+    //
+    stats.bytesLockMax = _memory.bytesMax();
+    stats.bytesLocked  = _memory.bytesLocked();
+    stats.numLocks     = _numLocks;
+    stats.numErrors    = _numErrors;
+    stats.numFiles     = MemFile::numFiles();
+
+    // The following requires a lock
+    //
+    hanMutex.lock();
+    stats.numFSets = hanCache.size();
+    hanMutex.unlock();
+    return stats;
+}
+
+/******************************************************************************/
+/*                             g e t S t a t u s                              */
+/******************************************************************************/
+
+MemMan::Status MemManReal::getStatus(Handle handle) {
+
+Status status;
+
+    // First check if this is a valid handle and, if so, find it in our cache.
+    // Once found, get its real status from the file set object.
+    //
+    if (handle != HandleType::INVALID && handle != HandleType::ISEMPTY)
+       {hanMutex.lock();
+        auto it = hanCache.find(handle);
+        if (it != hanCache.end() && it->second->isOwner(_memory))
+           {status = it->second->status();
+            hanMutex.unlock();
+            return status;
+           }
+        hanMutex.unlock();
+       }
+
+    // Return null status
+    //
+    memset(&status, 0, sizeof(status));
+    return status;
+}
+  
+/******************************************************************************/
+/*                                  l o c k                                   */
+/******************************************************************************/
+  
+MemMan::Handle MemManReal::lock(std::vector<TableInfo> const& tables, int chunk) {
+
+MemFileSet* fileSet;
+int         lockNum, flexNum, retc = 0;
+bool        mustLock;
+
+    // Make sure a table was specified
+    //
+    _numLocks++;
+    if (tables.size() == 0)
+       {_numErrors++;
+        errno = EINVAL;
+        return HandleType::INVALID;
+       }
```

Comment:
I'm not sure this empty table check is necessary. If this if statement is deleted this function will return HandleType::ISEMPTY, which should work without causing problems and means the caller wont have to make a separate code path for an empty list.